torch/csrc/autograd/variable.h

```diff
@@ -1,150 +1,255 @@
 #pragma once
 
-// A wrapper around at::Tensor to represent autograd Variables. Variables
-// can be implicitly converted to an at::Tensor.
+#include <Python.h>
 
-#include <mutex>
-#include <memory>
-#include <vector>
-#include <functional>
-#include <ATen/ATen.h>
-
-#include "torch/csrc/assertions.h"
-#include "torch/csrc/jit/ir.h"
-#include "torch/csrc/jit/tracer_state.h"
+#include "torch/csrc/autograd/edge.h"
 #include "torch/csrc/autograd/function_hook.h"
-#include "torch/csrc/utils/auto_unique_ptr.h"
 #include "torch/csrc/autograd/variable_version.h"
-#include "torch/csrc/autograd/edge.h"
-#include "torch/csrc/Types.h"
+#include "torch/csrc/jit/tracer_state.h"
+#include "torch/csrc/utils/auto_unique_ptr.h"
+
+#include <ATen/Scalar.h>
+#include <ATen/ScalarType.h>
+#include <ATen/Storage.h>
+#include <ATen/Tensor.h>
+#include <ATen/TensorImpl.h>
+#include <ATen/Type.h>
+
+#include <list>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <vector>
 
 namespace torch { namespace autograd {
 
-using at::Tensor;
-struct VariableImpl;
+struct Function;
+
+//===----------------------------------------------------------------------===//
+//                                Variable
+//===----------------------------------------------------------------------===//
+
+/// A `Variable` augments a `Tensor` with the ability to interact in our
+/// autograd machinery. `Variable` inherits from `Tensor` and may be converted
+/// to and from `Tensor` implicitly.
+class Variable : public at::Tensor {
```

Comment:
Why make it a `class` if it worked fine as `struct`, and was mentioned like this in so many places? `Function` is still a struct, so it will only be inconsistent.

Reply:
In most codebases `class` is used when `public` and `private` fields are used, to denote that this class/structure represents an abstraction, while `struct`s are used to signify a plain-old datatype, basically a pure data struct, with few methods and all fields exposed ([reference](https://stackoverflow.com/questions/54585/when-should-you-use-a-class-vs-a-struct-in-c)). I felt like using the same idiom in our codebase would be nice, and I was going to rename `Function` into a class next as I brush it up too. But I can change it to `struct`. What do you think?

Reply:
I'd rather stick to our current conventions instead of changing them. Having some types as `struct`s and other as `class`es is very annoying, because you have to remember which are which when forward declaring them, friending, etc. I'd rather keep them as `struct`s to be consistent